### PR TITLE
refactor(plugin-google-gtag, plugin-google-analytics): migrate packages to TS

### PR DESCRIPTION
**Pull request by Josh-Cena**
## Motivation

More migration, also provided typings to their `ThemeConfig` fields.

Hopefully this makes refactoring the `themeConfig` options to plugin options more convenient

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

### CODE REVIEW

File: packages/docusaurus-plugin-google-analytics/src/index.ts [hunk 1]
```diff
@@ -40,7 +43,7 @@ module.exports = function (context) {
       if (!isProd) {
         return {};
       }
-      return {
+      const HTMLTags = {
```
Comment (lex111) [inline packages/docusaurus-plugin-google-analytics/src/index.ts:46]: Don't understand, why create new const here, if we can return just object as it was before?

Comment (Josh-Cena) [reply packages/docusaurus-plugin-google-analytics/src/index.ts:46]: That doesn't work and spits out a giant "Type not assignable "error"... My best guess is due to some inference funkiness

File: packages/docusaurus-plugin-google-analytics/src/analytics.ts [hunk 2]
```diff
@@ -13,7 +13,7 @@ export default (function () {
   }
 
   return {
-    onRouteUpdate({location}) {
+    onRouteUpdate({location}: {location: Location}) {
```
Comment (slorber) [inline packages/docusaurus-plugin-google-analytics/src/analytics.ts:16]: We should probably create a type for js client modules

File: packages/docusaurus-preset-classic/src/preset-classic.d.ts [hunk 3]
```diff
@@ -14,10 +14,10 @@ export type Options = {
...
 
-export type ThemeConfig = import('@docusaurus/theme-common').ThemeConfig & {
-  // Those themeConfigs should rather be moved to preset/plugin options
-  // Plugin data can be made available to browser thank to the globalData api
-  algolia?: unknown; // TODO type plugin
-  googleAnalytics?: unknown; // TODO type plugin
-  gtag?: unknown; // TODO type plugin
-};
+export type ThemeConfig = import('@docusaurus/theme-common').ThemeConfig &
+  import('@docusaurus/plugin-google-analytics').ThemeConfig &
```
Comment (slorber) [inline packages/docusaurus-preset-classic/src/preset-classic.d.ts:18]: even for refactored TS packages, we still want to move those themeConfig to options. This feels overly complex to have to provide plugin in plugin array + config in theme config: configuring things in one place would provide better DX

**Review by slorber — APPROVED** [review]
LGTM thanks

added some comments but we can merge it

File: packages/docusaurus-preset-classic/src/preset-classic.d.ts (hunk 3 above)
Comment (Josh-Cena) [reply packages/docusaurus-preset-classic/src/preset-classic.d.ts:18]: We will move them to options in the future, this is just temporary

**Pull request merged**